# Assignment#7

**Name:** Lama Batta

**ID:** 1210922

<span style="color:blue"> **Import Libraries** </span>

In [1]:
import pandas as pd

**Describe the data set**

<span style= "color:red" > **Machine Learning Engineer Salary in 2024 Dataset from kaggle** </span>

**Feature Description**

- **work_year:** The year in which the salary data was collected .
- **experience_level:** The level of experience of the employee .
- **employment_type:** The type of employment ).
- **job_title:** The title of the job (e.g., Data Scientist).
- **salary:** The salary amount.
- **salary_currency:** The currency in which the salary is denominated.
- **salary_in_usd:** The salary amount converted to US Dollars.
- **employee_residence:** The country of residence of the employee.
- **remote_ratio:** The ratio indicating the level of remote work.
- **company_location:** The location of the company.
- **company_size:** The size of the company.

 **Load the data set**

In [2]:
# Read data from csv file
df = pd.read_csv('salaries.csv')
df

work_year experience_level employment_type                 job_title  \
0           2024               MI              FT            Data Scientist   
1           2024               MI              FT            Data Scientist   
2           2024               MI              CT            Data Scientist   
3           2024               MI              CT            Data Scientist   
4           2024               MI              FT      Data Science Manager   
...          ...              ...             ...                       ...   
16489       2020               SE              FT            Data Scientist   
16490       2021               MI              FT  Principal Data Scientist   
16491       2020               EN              FT            Data Scientist   
16492       2020               EN              CT     Business Data Analyst   
16493       2021               SE              FT      Data Science Manager   

        salary salary_currency  salary_in_usd employee_residence  \
0       120000             USD         120000                 AU   
1        70000             USD          70000                 AU   
2       130000             USD         130000                 US   
3       110000             USD         110000                 US   
4       240000             USD         240000                 US   
...        ...             ...            ...                ...   
16489   412000             USD         412000                 US   
16490   151000             USD         151000                 US   
16491   105000             USD         105000                 US   
16492   100000             USD         100000                 US   
16493  7000000             INR          94665                 IN   

       remote_ratio company_location company_size  
0                 0               AU            S  
1                 0               AU            S  
2                 0               US            M  
3                 0               US            M  
4                 0               US            M  
...             ...              ...          ...  
16489           100               US            L  
16490           100               US            L  
16491           100               US            S  
16492           100               US            L  
16493            50               IN            L  

[16494 rows x 11 columns]

**link kaggle for dataset**

<span style= "color:blue">https://www.kaggle.com/datasets/chopper53/machine-learning-engineer-salary-in-2024</span>

**The dictionary of the columns** 

In [3]:
# Data Type of columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16494 entries, 0 to 16493
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           16494 non-null  int64 
 1   experience_level    16494 non-null  object
 2   employment_type     16494 non-null  object
 3   job_title           16494 non-null  object
 4   salary              16494 non-null  int64 
 5   salary_currency     16494 non-null  object
 6   salary_in_usd       16494 non-null  int64 
 7   employee_residence  16494 non-null  object
 8   remote_ratio        16494 non-null  int64 
 9   company_location    16494 non-null  object
 10  company_size        16494 non-null  object
dtypes: int64(4), object(7)
memory usage: 1.4+ MB


<span style="color:blue">This result determines the data type for each column integer type  each of the  work_year,salary,salary_in_usd,and remote_ratio features,String type each of the experience_level, employment_type,job_title, salary_currency,employee_residence,company_location,and company_size </span>

**Data Cleaning**

In [37]:
Columns_drop={"company_location","remote_ratio","employee_residence"}
df1=df.drop(columns=Columns_drop)
df1

work_year experience_level employment_type                 job_title  \
0           2024               MI              FT            Data Scientist   
1           2024               MI              FT            Data Scientist   
2           2024               MI              CT            Data Scientist   
3           2024               MI              CT            Data Scientist   
4           2024               MI              FT      Data Science Manager   
...          ...              ...             ...                       ...   
16489       2020               SE              FT            Data Scientist   
16490       2021               MI              FT  Principal Data Scientist   
16491       2020               EN              FT            Data Scientist   
16492       2020               EN              CT     Business Data Analyst   
16493       2021               SE              FT      Data Science Manager   

        salary salary_currency  salary_in_usd company_size  
0       120000             USD         120000            S  
1        70000             USD          70000            S  
2       130000             USD         130000            M  
3       110000             USD         110000            M  
4       240000             USD         240000            M  
...        ...             ...            ...          ...  
16489   412000             USD         412000            L  
16490   151000             USD         151000            L  
16491   105000             USD         105000            S  
16492   100000             USD         100000            L  
16493  7000000             INR          94665            L  

[16494 rows x 8 columns]

 <span style="color:blue">**Dropping some columns** </span>

In [38]:
#cheak the missing value 
df1.isna().sum()

work_year           0
experience_level    0
employment_type     0
job_title           0
salary              0
salary_currency     0
salary_in_usd       0
company_size        0
dtype: int64

<span style="color:blue">The result indicates that there are no missing values for each features </span>

In [39]:
#Check for duplicates
df1.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
16489    False
16490    False
16491    False
16492    False
16493    False
Length: 16494, dtype: bool

<span style="color:blue">The result indicates that there are no duplicated values </span>

**Select Target and Features use to make the prediction**

In [40]:
# The target we are trying to predict
y = df1['salary']
# The features we will use to make the prediction
X = df1.drop(columns = 'salary')
X

work_year experience_level employment_type                 job_title  \
0           2024               MI              FT            Data Scientist   
1           2024               MI              FT            Data Scientist   
2           2024               MI              CT            Data Scientist   
3           2024               MI              CT            Data Scientist   
4           2024               MI              FT      Data Science Manager   
...          ...              ...             ...                       ...   
16489       2020               SE              FT            Data Scientist   
16490       2021               MI              FT  Principal Data Scientist   
16491       2020               EN              FT            Data Scientist   
16492       2020               EN              CT     Business Data Analyst   
16493       2021               SE              FT      Data Science Manager   

      salary_currency  salary_in_usd company_size  
0                 USD         120000            S  
1                 USD          70000            S  
2                 USD         130000            M  
3                 USD         110000            M  
4                 USD         240000            M  
...               ...            ...          ...  
16489             USD         412000            L  
16490             USD         151000            L  
16491             USD         105000            S  
16492             USD         100000            L  
16493             INR          94665            L  

[16494 rows x 7 columns]

**Split data to train and test**

In [41]:
# Import the train and test split from sklearn
from sklearn.model_selection import train_test_split

# Train test split 
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2,random_state=1210922)
X_train

work_year experience_level employment_type                  job_title  \
15690       2022               SE              FT             Data Scientist   
6564        2023               MI              FT             Data Scientist   
11669       2023               SE              FT             Data Scientist   
10313       2023               SE              FT              Data Engineer   
2466        2024               SE              FT             Data Architect   
...          ...              ...             ...                        ...   
16226       2022               SE              FT     Applied Data Scientist   
9389        2023               SE              FT               Data Analyst   
4156        2024               SE              FT  Machine Learning Engineer   
1210        2024               SE              FT             Data Architect   
10950       2023               EX              FT              Data Engineer   

      salary_currency  salary_in_usd company_size  
15690             USD         141525            M  
6564              USD          86600            M  
11669             USD         359170            M  
10313             USD         125000            M  
2466              USD         242100            M  
...               ...            ...          ...  
16226             USD         177000            L  
9389              GBP          61520            M  
4156              USD         324000            M  
1210              USD         108200            M  
10950             USD          71908            M  

[13195 rows x 7 columns]

<span style="color:blue">This result is after splitting of the data and using 80% train, making 13195 columns for X_train </span>

In [42]:
X_test

work_year experience_level employment_type         job_title  \
5915        2024               SE              FT  Research Analyst   
4408        2024               SE              FT      Data Analyst   
3038        2024               SE              FT     Data Engineer   
4692        2024               SE              FT      Data Analyst   
4966        2024               SE              FT    Data Scientist   
...          ...              ...             ...               ...   
16127       2022               SE              FT     Data Engineer   
2860        2024               EN              FT  Research Analyst   
6563        2023               MI              FT    Data Scientist   
8331        2023               SE              FT     Data Engineer   
8183        2023               SE              FT    Data Scientist   

      salary_currency  salary_in_usd company_size  
5915              USD         150000            M  
4408              USD         144700            M  
3038              USD         130000            M  
4692              USD         120000            M  
4966              USD         162000            M  
...               ...            ...          ...  
16127             USD         154000            M  
2860              USD         300000            M  
6563              USD         119200            M  
8331              USD          68400            M  
8183              USD          81500            M  

[3299 rows x 7 columns]

<span style="color:blue">This result is after splitting of the data and using 20% test, making 3299 columns for X_test </span>

In [43]:
numrical_columns = X_train.select_dtypes("number").columns
ordinal_columns = ['company_size']
nominal_columns = X_train.select_dtypes("object").drop(columns=ordinal_columns).columns


**Numrical pipeline**

In [44]:
#import the SimpleImputer ,make_pipeline and StandardScaler from  sklearn
from sklearn.impute import SimpleImputer 
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
# instantiate preprocessors
impute_median = SimpleImputer(strategy='median')
scaler = StandardScaler()

# Make a numeric preprocessing pipeline
numrical_pipe = make_pipeline(impute_median, scaler)
numrical_pipe

Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='median')),
                ('standardscaler', StandardScaler())])

**Ordinal pipline**

In [45]:

impute_na_ordinal = SimpleImputer(strategy='constant', fill_value='NA')

# Specifying the order of categories in quality/condition columns
company_size_order = ["S", "M", "L"]
# Making the list of order lists for OrdinalEncoder
ordinal_category_orders = [company_size_order]
# Instantiate the encoder and include the list of ordered values as an argument
ordinal_encoder = OrdinalEncoder(categories=ordinal_category_orders)

# Making a final scaler to scale category #'s
scaler_ord = StandardScaler()

ordinal_pipe = make_pipeline(impute_na_ordinal, ordinal_encoder, scaler_ord)

ordinal_pipe

Pipeline(steps=[('simpleimputer',
                 SimpleImputer(fill_value='NA', strategy='constant')),
                ('ordinalencoder',
                 OrdinalEncoder(categories=[['S', 'M', 'L']])),
                ('standardscaler', StandardScaler())])

**Nominal pipline**

In [46]:
impute_na_nominal = SimpleImputer(strategy='constant', fill_value = "NA")

# Instantiate one hot encoder
Nominal_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

# Make pipeline with imputer and encoder
Nominal_pipe = make_pipeline(impute_na_nominal,Nominal_encoder )
Nominal_pipe

Pipeline(steps=[('simpleimputer',
                 SimpleImputer(fill_value='NA', strategy='constant')),
                ('onehotencoder',
                 OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

**Build pipelines tuples**

In [47]:
from sklearn import set_config
set_config(transform_output='pandas')
numrical_tuple = ('numeric', numrical_pipe, numrical_columns)
ordinal_tuple = ('ordinal', ordinal_pipe, ordinal_columns)
Nominal_tuple = ('categorical', Nominal_pipe, nominal_columns)


# Create ColumnTransformer
# Instantiate with verbose_feature_names_out=False 
col_transformer = ColumnTransformer([numrical_tuple, ordinal_tuple, Nominal_tuple],
                                    verbose_feature_names_out=False)

# Fit on training data
col_transformer.fit(X_train)
# Transform the training data
X_train_transformer = col_transformer.transform(X_train)
# Transform the testing data
X_test_transformer = col_transformer.transform(X_test)
# View the processed training data
X_train_transformer



work_year  salary_in_usd  company_size  experience_level_EN  \
15690  -1.725631      -0.123245     -0.189427                  0.0   
6564   -0.319920      -0.909508     -0.189427                  0.0   
11669  -0.319920       2.992390     -0.189427                  0.0   
10313  -0.319920      -0.359804     -0.189427                  0.0   
2466    1.085790       1.316508     -0.189427                  0.0   
...          ...            ...           ...                  ...   
16226  -1.725631       0.384587      3.530047                  0.0   
9389   -0.319920      -1.268534     -0.189427                  0.0   
4156    1.085790       2.488924     -0.189427                  0.0   
1210    1.085790      -0.600300     -0.189427                  0.0   
10950  -0.319920      -1.119827     -0.189427                  0.0   

       experience_level_EX  experience_level_MI  experience_level_SE  \
15690                  0.0                  0.0                  1.0   
6564                   0.0                  1.0                  0.0   
11669                  0.0                  0.0                  1.0   
10313                  0.0                  0.0                  1.0   
2466                   0.0                  0.0                  1.0   
...                    ...                  ...                  ...   
16226                  0.0                  0.0                  1.0   
9389                   0.0                  0.0                  1.0   
4156                   0.0                  0.0                  1.0   
1210                   0.0                  0.0                  1.0   
10950                  1.0                  0.0                  0.0   

       employment_type_CT  employment_type_FL  employment_type_FT  ...  \
15690                 0.0                 0.0                 1.0  ...   
6564                  0.0                 0.0                 1.0  ...   
11669                 0.0                 0.0                 1.0  ...   
10313                 0.0                 0.0                 1.0  ...   
2466                  0.0                 0.0                 1.0  ...   
...                   ...                 ...                 ...  ...   
16226                 0.0                 0.0                 1.0  ...   
9389                  0.0                 0.0                 1.0  ...   
4156                  0.0                 0.0                 1.0  ...   
1210                  0.0                 0.0                 1.0  ...   
10950                 0.0                 0.0                 1.0  ...   

       salary_currency_JPY  salary_currency_MXN  salary_currency_NOK  \
15690                  0.0                  0.0                  0.0   
6564                   0.0                  0.0                  0.0   
11669                  0.0                  0.0                  0.0   
10313                  0.0                  0.0                  0.0   
2466                   0.0                  0.0                  0.0   
...                    ...                  ...                  ...   
16226                  0.0                  0.0                  0.0   
9389                   0.0                  0.0                  0.0   
4156                   0.0                  0.0                  0.0   
1210                   0.0                  0.0                  0.0   
10950                  0.0                  0.0                  0.0   

       salary_currency_PHP  salary_currency_PLN  salary_currency_SGD  \
15690                  0.0                  0.0                  0.0   
6564                   0.0                  0.0                  0.0   
11669                  0.0                  0.0                  0.0   
10313                  0.0                  0.0                  0.0   
2466                   0.0                  0.0                  0.0   
...                    ...                  ...                  ...   
16226                  0.0                 

<span style="color:e">**Fit a linear regression model on the transformed data.** </span>

In [48]:
from sklearn.linear_model import LinearRegression
# from sklearn.linear_model SGDRegressor
model = LinearRegression()
model_pipe = make_pipeline(col_transformer, model)
# Fit the model on the training data
model_pipe.fit(X_train, y_train)
model_pipe

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('numeric',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('standardscaler',
                                                                   StandardScaler())]),
                                                  Index(['work_year', 'salary_in_usd'], dtype='object')),
                                                 ('ordinal',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value='NA',
                                                                                 strategy='constant')),
                                                                  ('ordinalencoder...
                                                  ['company_size']),
                                                 ('categorical',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value='NA',
                                                                                 strategy='constant')),
                                                                  ('onehotencoder',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  Index(['experience_level', 'employment_type', 'job_title', 'salary_currency'], dtype='object'))],
                                   verbose_feature_names_out=False)),
                ('linearregression', LinearRegression())])

<span style="color:red">**Print the model parameters and provide an interpretation for each.** </span>

In [49]:
b_0 = model.intercept_.round(2)
b_i = model.coef_.round(2)
feature_names = col_transformer.get_feature_names_out()
eq = " +\n".join([ f"{'' if i == 0 else ' '*24}{b_i[i]} * {feature_names[i]}"  for i in range(len(feature_names))])
print(f"Model equation: y_hat = {eq}")

Model equation: y_hat = 667.28 * work_year +
                        69707.91 * salary_in_usd +
                        4196.81 * company_size +
                        4363015412504446.5 * experience_level_EN +
                        4363015412503638.0 * experience_level_EX +
                        4363015412507001.5 * experience_level_MI +
                        4363015412511611.5 * experience_level_SE +
                        723973630295157.2 * employment_type_CT +
                        723973630269081.5 * employment_type_FL +
                        723973630267254.4 * employment_type_FT +
                        723973630274120.0 * employment_type_PT +
                        -6509529641227.82 * job_title_AI Architect +
                        -6509529635229.92 * job_title_AI Developer +
                        -6509529643466.95 * job_title_AI Engineer +
                        -6509529645996.16 * job_title_AI Product Manager +
                        -6509529264569.9 * job

**Interpret**

- **667.28 (work_year):** 

 <span style="color:blue">When the work year incease by one year, the salary expected to increase by 667.28. </span>

- **69707.91 (salary_in_usd):**

 <span style="color:blue">When the salary_in_usd incease by one usd , the salary expected to increase by 69707.91. </span>

- **4196.81 (company_size):**

<span style="color:blue">When the company_size incease by one size , the salary expected to increase by 4196.81. </span>

-  **4363015412504446.5(experience_level_EN):**

<span style="color:blue">When the increases experience_level_EN by one level ,the salary expected to increases by 4363015412504446.5. </span>

-  **4363015412503638.0(experience_level_EX):**

<span style="color:blue">When the increases experience_level_Ex by one level ,the salary expected to increases by 4363015412503638.0. </span>

-  **4363015412507001.5(experience_level_MI):**

<span style="color:blue">When the increases experience_level_MI by one level ,the salary expected to increases by  4363015412507001.5. </span>

-  **4363015412511611.5(experience_level_SE):**

<span style="color:blue">When the increases experience_level_SE by one level ,the salary expected to increases by 4363015412511611.5. </span>

- **723973630295157.2 (employment_type_CT)** 

<span style="color:blue">When the employment_type change to CT ,the salary expected to increases by 723973630295157.2. </span>

- **723973630269081.5 (employment_type_FL)** 

<span style="color:blue">When the employment_type change to FL ,the salary expected to increases by  723973630269081.5. </span>

- **723973630267254.4 (employment_type_FT)**

<span style="color:blue">When the employment_type change to FT ,the salary expected to increases by 723973630267254.4. </span>

- **723973630274120.0. (employment_type_PT)**

<span style="color:blue">When the employment_type change to PT ,the salary expected to increases by 723973630274120.0. </span>

<span style="color:red">**Evaluate the model's quality using appropriate metrics.** </span>

In [17]:
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
def regression_metrics(y_true, y_pred, label='', verbose = True, output_dict=False):
    # Get metrics
    mae = mean_absolute_error(y_true, y_pred)
    mse = mean_squared_error(y_true, y_pred)
    rmse = mean_squared_error(y_true, y_pred, squared=False) 
    r_squared = r2_score(y_true, y_pred)
    if verbose == True:
        # Print Result with Label and Header
        header = "-"*60
        print(header, f"Regression Metrics: {label}", header, sep='\n')
        print(f"- MAE = {mae:,.3f}")
        print(f"- MSE = {mse:,.3f}")
        print(f"- RMSE = {rmse:,.3f}")
        print(f"- R^2 = {r_squared:,.3f}")
    if output_dict == True:
        metrics = {'Label':label, 'MAE':mae,
                 'MSE':mse, 'RMSE':rmse, 'R^2':r_squared}
        
        
# Get predictions for training data
y_train_pred = model_pipe.predict(X_train)

# Call the helper function to obtain regression metrics for training data
results_train = regression_metrics(y_train, y_train_pred, label='Training Data')
print()
# Get predictions for test data
y_test_pred = model_pipe.predict(X_test)
# Call the helper function to obtain regression metrics for test data
results_test = regression_metrics(y_test, y_test_pred, label='Test Data' )

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 10,042.939
- MSE = 6,826,010,990.508
- RMSE = 82,619.677
- R^2 = 0.943

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 10,982,907,285,728.592
- MSE = 20,902,771,842,190,501,550,240,038,912.000
- RMSE = 144,577,909,246,850.375
- R^2 = -203,767,075,040,839,232.000


C:\Users\user\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(
C:\Users\user\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


<span style="color:red">**Determine if the model exhibits overfitting or underfitting and explain your reasoning.** </span>

<span style="color:blue">According to the results that appeared, the value of RMSE in Training Data is 82,619.677 and its value in Test Data has become 144,577,909,246,850.375. This means that it is increasing, so the model is overfitting. </span>